src/ui/camera.js

```diff
@@ -454,22 +446,32 @@ class Camera extends Evented {
         const upperRight = new Point(Math.max(p0rotated.x, p1rotated.x), Math.max(p0rotated.y, p1rotated.y));
         const lowerLeft = new Point(Math.min(p0rotated.x, p1rotated.x), Math.min(p0rotated.y, p1rotated.y));
 
-        const offset = Point.convert(options.offset),
-            size = upperRight.sub(lowerLeft),
-            scaleX = (tr.width - lateralPadding * 2 - Math.abs(offset.x) * 2) / size.x,
-            scaleY = (tr.height - verticalPadding * 2 - Math.abs(offset.y) * 2) / size.y;
+        // Calculate zoom: consider the original bbox and padding.
+        const size = upperRight.sub(lowerLeft);
+        const scaleX = (tr.width - options.padding.left - options.padding.right) / size.x;
+        const scaleY = (tr.height - options.padding.top - options.padding.bottom) / size.y;
 
         if (scaleY < 0 || scaleX < 0) {
             warnOnce(
                 'Map cannot fit within canvas with the given bounds, padding, and/or offset.'
             );
             return;
         }
-        options.center =  tr.unproject(p0world.add(p1world).div(2));
-        options.zoom = Math.min(tr.scaleZoom(tr.scale * Math.min(scaleX, scaleY)), options.maxZoom);
-        options.bearing = bearing;
 
-        return options;
+        const zoom = Math.min(tr.scaleZoom(tr.scale * Math.min(scaleX, scaleY)), options.maxZoom);
+
+        // Calculate center: apply the zoom, the configured offset, as well as offset that exists as a result of padding.
+        const paddingOffset = [(options.padding.left - options.padding.right) / 2, (options.padding.top - options.padding.bottom) / 2];
+        const offsetAtInitialZoom = Point.convert([options.offset[0] + paddingOffset[0], options.offset[1] + paddingOffset[1]]);
```

Comment:
Nitpick: let's avoid temporary arrays here by using `new Point(x, y)` instead of `Point.convert([x, y])`, and maybe declaring separate `paddingOffsetX` and `paddingOffsetY` variables since we don't really need it being in an array.

Reply:
Also, `options.offset` is of `PointLike` type so this code will break if you pass e.g. `offset: new Point(10, 10)` instead of `offset: [10, 10]`. So let's bring back the `const offset = Point.convert(options.offset)` line and use that.

Reply:
Great points! It looks like the previous implementation was also accessing array indices directly, so I added a test to make sure `PointLike` objects work.